### ISSUE TITLE
Prettier changes block comment node location in AST

### ISSUE THREAD

**Issue by TheSavior** [prettier/prettier#4398]
**Prettier 1.12.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEAxCFMC8mwmAOlJpZgEYCGATgBQCUxAvgDSYD0AVFQMFDhI0WPESJ5Ab27SqdAF4t25NgG4QHEBAAOMAJbpkoBvQgB3AAoMEaZCFoA3CAYAmWkNXq0wAazgYAGVdXwMoAHNkGHoAVzhtcLQ4ehgrHwiAW1pkADNaABtk7QArNAAPACEff0Cg2ky4ABlwuDzC4pBQ+mT6BzpqAE8C6E9denCYAHV3GAALZAAOAAZtcYhkqZ9dB3G4Xqc27Xo4AEdYgxP02iycpHyihJBkzINouKe0cIiCuABFWIQeDtR7aGC0agzNzzZAAJjBPgMBW+AGEIJlsg4oNAjiBYskACoQ+z3DpwNhsIA)
```sh
# Options (if any):
--single-quote
```

**Input:**
```jsx
const Foo = { 
    bar() {}, /*                                                                                       
    */
    baz() {}
};```

**Output:**
```jsx
const Foo = {
  bar() {} /*                                                                                       
    */, // Note this trailing comma is after the comment instead of before
  baz() {}
};
```

**Expected behavior:**
```
const Foo = { 
    bar() {}, 
    /*                                                                                       
    */
    baz() {}
};
```

In my input code, the comment is seen by the ast as a leadingComment to baz and **not** a trailing comma to bar. The resulting output from prettier is seen by the ast as having the comment be both a trailing comment to bar and a leading comment to baz (I find this part super surprising because of the comma, but that is an aside).

Overall, prettier shouldn't move the comment to be on a different side of the comma.

**Comment by lydell** [comment]
Comments are super difficult to get right in Prettier, so if it ends up in a reasonable place we usually think it’s good enough. Could you share your _real_ use case so we know how and why to address this?

**Comment by TheSavior** [comment]
I’m building a codegen at work that creates a bunch of files where each object method is a test case that a developer has to fill in and has an associated docblock the codegen creates that another script scrapes to associate meta information with the test.

I am currently working around this issue by taking the output of my codegen (Babel-generator), running a regex replace on the output string to add blank lines before the comments, and then passing it to prettier before writing it to the file system.

**Comment by lydell** [comment]
Can you show some sample code? (The code in the OP does not look like real sample code to me.)

Generally it has worked best to put magic comments on the line above the line it relates to. For example, we usually recommend `// eslint-diable-next-line` over `//eslint-disable-line`. Since you are building the codegen yourself, is this something you can take into account? Not saying there's nothing that can be done about the comment placement here, but since you seem to have full control I'd say don't make things harder for yourself :)

**Comment by TheSavior** [comment]
Something worth noting is that prettier isn't running over an existing file that I can modify. Prettier is being used to consistently *format* a code generated file. The output file is *entirely* created by constructing an ast with babel utilities and then formatting that result with prettier.

The codegen script creates output like this:
```
const Tests = {
  /**
   * @uidocs FIGButton
   * @design https://fburl.com/blahblah
   */
  testOutlineWhiteMediumIconLabelDefault() {
    return <ServerSnapshot />;
  },
  /**
   * @uidocs FIGButton
   * @design https://fburl.com/blahblah2
   */
  testOutlineWhiteMediumIconLabelDisabled() {
    return <ServerSnapshot />;
  },
};
```

My codegen outputs this file by doing the following steps:

```
const generated = babelGenerate(newAst, { retainLines: true }, initialCode).code;

const fixup = generated
  // Handle prettier moving docblocks to before the comma. 
  // It doesn't do that if we add a new line
  .replace(/},\s\/\*\*/g, '},\n/**') 
  // Add a blank line after the leading comment so that prettier puts it on the line above
  // instead of on the same line as the object method
  .replace(/\*\/(\s)*test/g, '*/\ntest');
return prettier.format(fixup, standardPrettierOptions);
```

**Comment by lydell** [comment]
Wow, that’s a very specific use case. Just one more thing: Can you show what the code from `babelGenerate` looks like (before running Prettier)?

(Related: #465)

**Comment by TheSavior** [comment]
It looks like the input in my initial post:
```
const Tests = {
  /**
   * @uidocs FIGButton
   * @design https://fburl.com/blahblah
   */
  testOutlineWhiteMediumIconLabelDefault() {
    return <ServerSnapshot />;
  }, /**        // Regex #1 moves this comment to start on the next line
   * @uidocs FIGButton
   * @design https://fburl.com/blahblah2
   */ testOutlineWhiteMediumIconLabelDisabled() {     // Regex #2 moves this test to start on the next line
    return <ServerSnapshot />;
  },
};
```

**Comment by lydell** [comment]
I'm really confused now – I can't seem to reproduce this on the playground?

**Prettier 1.12.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEAVOG2YC8mwAOlJpgPQBUN5lmNmAAgK4CWAJhGAQGIBJAOIAhNjBjQGlZiy54OAcwoALSQAc0SKlQBmAIzYAnADYA6SAFsqB0wENVdxzKZVM8DAHkJpjlDgAdVUOeABZOC4ONitBSCgAGXsDOFMAETg9ezZTGAAKAEoSXUwAJTglOAAPTABiACZMKwgANzwPEIJPLClMDHtjLGgOuEwAqqw-ANdMYzgYEwoAHgBlOGM24xWoey1VCCwqAD4AbhkAXwAaajpGO-uHx8eS8sqa2oBGJtb2mE7MaxWBA9CB9GADIYUP6jcaTfxwVxyTg8PiYIRiCRSKCI1gKNDKNSabS6QwmCzWWwOJxU+qI9zdHwwKZBELhSLRWLxJIpdIcNDJUyRQokRgzOYLYzLNYbdbbXZofaHU4XS7kc4nECXEAQDQwDjoZCgAbGCAAdwACgMEGhkCB7C0INxNSADMZ7GAANbzFYad3+JTIGDGNhwLX+NDrGDmt1KKz2ZBZUwRrUAKzQVVEbs93vsQIS8IT9iToZAvuMEeMtoMyQAnqZoM6NMZ-DBAtw-sgABwABi1TYgEcCbo0tqbeHWbWdcwAjpw5tH7LH40hE8mQBGrBxA8GS-ioEpBQBFNgHOCF4ta8EGNtcDtIeqXt0cKZKADCECscdtUGgZ61bAjbBkhtFciwjc5ziAA)

**Input:**
```jsx
const Tests = {
  /**
   * @uidocs FIGButton
   * @design https://fburl.com/blahblah
   */ testOutlineWhiteMediumIconLabelDefault() {// Regex #2 moves this test to start on the next line
    return <ServerSnapshot />;
  }, /**                                // Regex #1 moves this comment to start on the next line
   * @uidocs FIGButton
   * @design https://fburl.com/blahblah2
   */ testOutlineWhiteMediumIconLabelDisabled() {   
    return <ServerSnapshot />;
  },
};
```

**Output:**
```jsx
const Tests = {
  /**
   * @uidocs FIGButton
   * @design https://fburl.com/blahblah
   */ testOutlineWhiteMediumIconLabelDefault() {
    // Regex #2 moves this test to start on the next line
    return <ServerSnapshot />;
  },
  /**                                // Regex #1 moves this comment to start on the next line
   * @uidocs FIGButton
   * @design https://fburl.com/blahblah2
   */ testOutlineWhiteMediumIconLabelDisabled() {
    return <ServerSnapshot />;
  }
};

```

**Comment by duailibe** [comment]
Yeah there's a small difference between the OP and the real code you posted, the 2nd function is on the same line as the end of the comment, not on a new line. See the following example:

**Prettier 1.12.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEAxCFMC8mwmAOlJpZgEYCGATgBQCUxAvgDSYD0AVFQMFDhI0WPESJ5Ab27SqdAF4t25NgG5y5SFAw48hYmQoKGK4Jx79JN23fvj5lWTVrLWFtepAcQEAA4wAJboyKAM9BAA7gAKDAhoyCC0AG4QQQAmPiDU9LRgANZwMADK-vlBUADmyDD0AK5wvpVocPQwMXlVALa0yABmtAA2rb4AVmgAHgBCeYXFJbTdcAAylXADw6Mg5fSt9El01ACeQ9DZ-vSVMADqmTAAFsgAHAAMvpcQrTd5-kmXcH2KQ2vnocAAjvUgmDOrQen0kIMRk0QK1ukFag0UWhKlUhnAAIr1CDwTbI3wwWjUO4ZR7IABMFLyQSGuIAwhBur0klBoCCQPVWgAVKmJRFbOBsNhAA)

**Input:**
```jsx
const Foo = { 
    bar() {}, /*                                                                                       
    */
    baz() {}
};

const Foo = { 
    bar() {}, /*                                                                                       
    */ baz() {}
};
```

**Output:**
```jsx
const Foo = {
  bar() {} /*                                                                                       
    */,
  baz() {}
};

const Foo = {
  bar() {},
  /*                                                                                       
    */ baz() {}
};

```

In the 1st case, prettier considers the comment a trailing comment of `bar()` and so puts the comment before the comma. In the 2nd case, it considers the comment as leading comment of `baz()` and so puts the comment after the comma.

As @lydell said, it's tricky to get it right and there certainly are a lot of edge cases we haven't seen yet.

**Comment by TheSavior** [comment]
@duailibe Thanks for tracking down that subtle difference in what I posted 👍

**Comment by mbrookes** [comment]
After disabling no-unexpected-multiline as recommended in the docs, prettier reformats the code such that the error comes back:

```diff
         };
       };
     })
-    // eslint-disable-next-line no-unexpected-multiline
-    ["catch"](function(error) {
+    [
+      // eslint-disable-next-line no-unexpected-multiline
+      "catch"
+    ](function(error) {
       console.error("Error during service worker registration:", error);
     });
 }
```

**Comment by lydell** [comment]
@mbrookes That’s not related to this issue, unfortunately.

For others reading this, @mbrookes comment is a cross-post of https://github.com/prettier/eslint-config-prettier/issues/32#issuecomment-531874121, for context.

**Comment by mbrookes** [comment]
Oops, yes, sorry, block comments here I see. I was looking for a relevant issue that wasn't already closed.

**Comment by supersnager** [comment]
HI, i have also problem with this. And i have unusual use case.
I'm using prettier in browser to format code. In my project i'm using some tokens in comments e.g. ``` /*MY_TOKEN*/```
Some of this tokens are after the comma:

```javascrtipt
{
  foo: "foo",/*MY_TOKEN*/
  bar: "bar
}
```

After prettier i'm doing some processing on generated pretty code and it's important for me to have ```/*MY_TOKEN*/``` where it should be - that is **after** comma :)

**Comment by brodybits** [comment]
I think that anything that changes the AST like this should be called a bug.

**Comment by thorn0** [comment]
No, it shouldn't. Comments, commas, new lines aren't considered part of the AST.
